src/encryption.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright The ocicrypt Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use std::collections::HashMap;
+use std::io::Read;
+
+use crate::config::DecryptConfig;
+use crate::config::EncryptConfig;
+
+use crate::blockcipher::{
+    Finalizer, LayerBlockCipherHandler, LayerBlockCipherOptions, PrivateLayerBlockCipherOptions,
+    PublicLayerBlockCipherOptions, AES256CTR,
+};
+use crate::keywrap::jwe::JweKeyWrapper;
+use crate::keywrap::KeyWrapper;
+extern crate oci_distribution;
+use oci_distribution::manifest::OciDescriptor;
+extern crate base64;
+
+lazy_static! {
+    static ref KEY_WRAPPERS: HashMap<String, Box<dyn KeyWrapper>> = {
+        let mut m = HashMap::new();
+        m.insert(
+            "jwe".to_string(),
+            Box::new(JweKeyWrapper {}) as Box<dyn KeyWrapper>,
+        );
+        m
+    };
+    static ref KEY_WRAPPERS_ANNOTATIONS: HashMap<String, String> = {
+        let mut m = HashMap::new();
+        for (scheme, key_wrapper) in KEY_WRAPPERS.iter() {
+            m.insert(key_wrapper.annotation_id().to_string(), scheme.clone());
+        }
+        m
+    };
+}
+
+// EncryptLayerFinalizer can get the annotations to set for the encrypted layer
+#[derive(Debug, Default, Clone)]
+pub struct EncLayerFinalizer {
+    lbco: LayerBlockCipherOptions,
```

Comment:
I'm not as familiar with rust, but from a very limited understanding, I am wondering if fields of the `finalized_annotations` can be put in the struct, in the go implementation they were just returned as a function closure that included all those fields. This would make it so that we don't have to expose the block encryption finalizer to the calling function of `encrypt_layer`..

So instead of 
```
        let new_annotations = elf
            .finalized_annotations(&ec, &desc, Some(&mut encryptor))
            .unwrap();
```
It could look something like
```
        let new_annotations = elf
            .finalized_annotations()
            .unwrap();
```


Not sure what's possible in rust, so please validate this for me. Thanks!

Reply:
In rust, return a closure which need bind to a struct describe at  : https://doc.rust-lang.org/book/ch19-05-advanced-functions-and-closures.html#returning-closures
It's a good suggestion to add these fields to the struct EncLayerFinalizer , but rust also have a data lifecycle and ownership restrict, add them may make code more complicate.